gax-httpjson/src/main/java/com/google/api/gax/httpjson/ApiMessageHttpRequestFormatter.java

```diff
@@ -64,20 +124,78 @@ public ApiMessageHttpRequestFormatter(ResourceNameFactory resourceNameFactory) {
   }
 
   @Override
-  public Map<String, String> getPathParams(T apiMessage, String resourceNameField) {
-    String resourceNamePath = apiMessage.getFieldStringValue(resourceNameField);
+  public String getRequestBody(ApiMessage apiMessage) {
+    ApiMessage body = apiMessage.getApiMessageRequestBody();
+    if (body != null) {
+      return getRequestMarshaller().toJson(body);
+    }
+    return null;
+  }
+
+  @Override
+  public String getPath(RequestT apiMessage) {
+    Map<String, String> pathParams = getPathParams(apiMessage);
+    return getEndpointPathTemplate().instantiate(pathParams);
+  }
+
+  private Map<String, String> getPathParams(RequestT apiMessage) {
+    String resourceNamePath = apiMessage.getFieldStringValue(getResourceNameField());
     if (resourceNamePath == null) {
       throw new IllegalArgumentException(
-          String.format("Resource name field %s is null in message object.", resourceNameField));
+          String.format(
+              "Resource name field %s is null in message object.", getResourceNameField()));
     }
-    return resourceNameFactory.parse(resourceNamePath).getFieldValuesMap();
+    return getResourceNameFactory().parse(resourceNamePath).getFieldValuesMap();
   }
 
-  @Override
-  public void writeRequestBody(ApiMessage apiMessage, Gson marshaller, Appendable writer) {
-    ApiMessage body = apiMessage.getApiMessageRequestBody();
-    if (body != null) {
-      marshaller.toJson(body, writer);
+  public static class Builder<RequestT extends ApiMessage> {
```

Comment:
I'm not sure how to use AutoValue.Builder when I need a constructor (or the static `create()`) to create a private field, the Gson object.

Reply:
You can override `build()` - for example: https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/batching/FlowControlSettings.java#L91

Reply:
Hmm I'm not a huge fan of doing that because right now there's a field in the getters that's not in the Builder setters, and there's a field in the Builder setters that's not in the getters. Doing something like in `FlowControlSettings` would require me to expose a getter and setter that don't need to be there, because the set of setter fields and the set of getter fields would have to be the same.